File: src/main/java/org/dimdev/dimdoors/DimensionalDoorsPlugin.java
package org.dimdev.dimdoors;

import com.hypixel.hytale.event.EventRegistry;
import com.hypixel.hytale.logger.HytaleLogger;
import com.hypixel.hytale.server.core.plugin.JavaPlugin;
import com.hypixel.hytale.server.core.plugin.JavaPluginInit;
import org.dimdev.dimdoors.commands.DimensionalDoorsPluginCommand;
import org.dimdev.dimdoors.systems.FabricOfRealityPlacementSystem;

import java.util.logging.Level;

/**
 * DimensionalDoors - A Hytale server plugin.
 */
public class DimensionalDoorsPlugin extends JavaPlugin {
    private static final HytaleLogger LOGGER = HytaleLogger.forEnclosingClass();

    private static DimensionalDoorsPlugin instance;

    public DimensionalDoorsPlugin(JavaPluginInit init) {
        super(init);
        instance = this;
    }

    public static DimensionalDoorsPlugin getInstance() {
        return instance;
    }

    @Override
    protected void setup() {
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Setting up...");

        registerCommands();
        registerListeners();
        registerSystems();

        LOGGER.at(Level.INFO).log("[DimensionalDoors] Setup complete!");
    }

    private void registerCommands() {
        try {
            getCommandRegistry().registerCommand(new DimensionalDoorsPluginCommand());
            LOGGER.at(Level.INFO).log("[DimensionalDoors] Registered /dimdoors command");
        } catch (Exception e) {
            LOGGER.at(Level.WARNING).withCause(e).log("[DimensionalDoors] Failed to register commands");
        }
    }

    private void registerListeners() {
    }

    private void registerSystems() {
        try {
            getEntityStoreRegistry().registerSystem(new FabricOfRealityPlacementSystem());
            LOGGER.at(Level.INFO).log("[DimensionalDoors] Registered Fabric of Reality placement system");
        } catch (Exception e) {
            LOGGER.at(Level.WARNING).withCause(e).log("[DimensionalDoors] Failed to register ECS systems");
        }
    }

    @Override
    protected void start() {
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Started!");
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Use /dimdoors help for commands");
    }

    @Override
    protected void shutdown() {
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Shutting down...");
        instance = null;
    }
}

File: src/main/kotlin/org/dimdev/dimdoors/DimensionalDoorsPlugin.kt
package org.dimdev.dimdoors

import com.hypixel.hytale.server.core.plugin.JavaPlugin
import com.hypixel.hytale.server.core.plugin.JavaPluginInit
import com.hypixel.hytale.event.EventRegistry
import com.hypixel.hytale.logger.HytaleLogger
import org.dimdev.dimdoors.commands.DimensionalDoorsPluginCommand
import org.dimdev.dimdoors.listeners.PlayerListener
import org.dimdev.dimdoors.systems.FabricOfRealityPlacementSystem
import java.util.logging.Level

/**
 * DimensionalDoors - A Hytale server plugin.
 */
class DimensionalDoorsPlugin(init: JavaPluginInit) : JavaPlugin(init) {

    companion object {
        private val LOGGER = HytaleLogger.forEnclosingClass()

        @JvmStatic
        var instance: DimensionalDoorsPlugin? = null
            private set
    }

    init {
        instance = this
    }

    override fun setup() {
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Setting up...")

        // Register commands
        registerCommands()

        // Register event listeners
        registerListeners()

        // Register ECS systems
        registerSystems()

        LOGGER.at(Level.INFO).log("[DimensionalDoors] Setup complete!")
    }

    private fun registerCommands() {
        try {
            commandRegistry.registerCommand(DimensionalDoorsPluginCommand())
            LOGGER.at(Level.INFO).log("[DimensionalDoors] Registered /dimdoors command")
        } catch (e: Exception) {
            LOGGER.at(Level.WARNING).withCause(e).log("[DimensionalDoors] Failed to register commands")
        }
    }

    private fun registerListeners() {
        val eventBus: EventRegistry = eventRegistry

        try {
            PlayerListener().register(eventBus)
            LOGGER.at(Level.INFO).log("[DimensionalDoors] Registered player event listeners")
        } catch (e: Exception) {
            LOGGER.at(Level.WARNING).withCause(e).log("[DimensionalDoors] Failed to register listeners")
        }
    }

    private fun registerSystems() {
        try {
            entityStoreRegistry.registerSystem(FabricOfRealityPlacementSystem())
            LOGGER.at(Level.INFO).log("[DimensionalDoors] Registered Fabric of Reality placement system")
        } catch (e: Exception) {
            LOGGER.at(Level.WARNING).withCause(e).log("[DimensionalDoors] Failed to register ECS systems")
        }
    }

    override fun start() {
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Started!")
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Use /dimdoors help for commands")
    }

    override fun shutdown() {
        LOGGER.at(Level.INFO).log("[DimensionalDoors] Shutting down...")
        instance = null
    }
}
